packages/hop-node/src/db/TransfersDb.ts

```diff
@@ -36,8 +42,61 @@ export type Transfer = {
 }
 
 class TransfersDb extends BaseDb {
-  async update (transferId: string, data: Partial<Transfer>) {
-    return super.update(transferId, data)
+  ready = false
+
+  constructor (prefix: string, _namespace?: string) {
+    super(prefix, _namespace)
+
+    // this only needs to be ran once on start up to backfill timestamped keys.
+    // this function can be removed once all bonders update.
+    // timestamped keys (in addition to transferId as keys) are needed to filter
+    // leveldb read streams.
+    this.trackTimestampedKeys()
+      .then(() => {
+        this.ready = true
+        this.logger.debug('transfersDb ready')
+      })
+      .catch(this.logger.error)
+  }
+
+  private async tilReady (): Promise<boolean> {
+    if (this.ready) {
+      return true
+    }
+
+    await wait(100)
+    return this.tilReady()
+  }
+
+  async trackTimestampedKeys () {
+    const transfers = await this.getTransfers()
+    for (const transfer of transfers) {
+      await this.trackTimestampedKey(transfer)
+    }
+  }
+
+  async trackTimestampedKey (transfer: Partial<Transfer>) {
+    const transferId = transfer.transferId
+    const key = this.getTimestampedKey(transfer)
+    if (!key || !transferId) {
+      return
+    }
+    const exists = await this.getById(key)
+    if (!exists) {
+      await super.update(key, { transferId })
```

Comment:
Will we have two entries per transfer (for all old txs) -- one with a timestamp and one without?

Reply:
correct. it's filtered here

![DeepinScreenshot_select-area_20211004125040](https://user-images.githubusercontent.com/168240/135915224-4dbbb13e-ad5c-4ada-bbd9-ee1d5a0b1206.png)

I'm looking into ways to use another subleveldb for the timestamped keys only

Reply:
understood 👍